cmake/test/gtest.cmake

```diff
@@ -85,97 +85,261 @@ function(catkin_add_executable_with_gtest target)
+          set(GTEST_FROM_SOURCE_LIBRARY_DIRS ${gtest_lib_dir} CACHE INTERNAL "")
+          set(GTEST_FROM_SOURCE_LIBRARIES ${gtest_libs} CACHE INTERNAL "")
+          set(GTEST_FROM_SOURCE_MAIN_LIBRARIES ${gtest_main_libs} CACHE INTERNAL "")
+
+          set(_CATKIN_SKIP_INSTALL_RULES TRUE)
+          function(install)
+            if(_CATKIN_SKIP_INSTALL_RULES)
+              return()
+            endif()
+            _install(${ARGN})
+          endfunction()
+          add_subdirectory(${gtest_base_dir} ${gtest_lib_dir})
+          set(_CATKIN_SKIP_INSTALL_RULES FALSE)
+          set_target_properties(${gtest_libs} ${gtest_main_libs}
+                                PROPERTIES EXCLUDE_FROM_ALL 1)
+
+          message(STATUS "Found gtest sources under '${gtest_base_dir}': gtests will be built")
+        else()
+          if(CATKIN_TOPLEVEL)
+            message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system or checkout gtest (by running 'svn checkout http://googletest.googlecode.com/svn/tags/release-1.6.0 gtest' in the source space '${CMAKE_SOURCE_DIR}' of your workspace) to enable gtests")
+          else()
+            message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system to enable gtests")
+          endif()
+        endif()
       endif()
     endif()
-  endif()
-  if(GTEST_FROM_SOURCE_FOUND)
-    # set the same variables as find_package()
-    # do NOT set GTEST_FOUND in the cache since when using gtest from source
-    # we must always add the subdirectory to have the gtest targets defined
-    set(GTEST_FOUND ${GTEST_FROM_SOURCE_FOUND})
-    set(GTEST_INCLUDE_DIRS ${GTEST_FROM_SOURCE_INCLUDE_DIRS})
-    set(GTEST_LIBRARY_DIRS ${GTEST_FROM_SOURCE_LIBRARY_DIRS})
-    set(GTEST_LIBRARIES ${GTEST_FROM_SOURCE_LIBRARIES})
-    set(GTEST_MAIN_LIBRARIES ${GTEST_FROM_SOURCE_MAIN_LIBRARIES})
-    set(GTEST_BOTH_LIBRARIES ${GTEST_LIBRARIES} ${GTEST_MAIN_LIBRARIES})
+
+    if(GMOCK_FROM_SOURCE_FOUND)
+      # set the same variables as find_package()
+      # do NOT set in the cache since when using gmock/gtest from source
+      # we must always add the subdirectory to have their targets defined
+      set(GMOCK_FOUND ${GMOCK_FROM_SOURCE_FOUND})
+      set(GMOCK_INCLUDE_DIRS ${GMOCK_FROM_SOURCE_INCLUDE_DIRS})
+      set(GMOCK_LIBRARY_DIRS ${GMOCK_FROM_SOURCE_LIBRARY_DIRS})
+      set(GMOCK_LIBRARIES ${GMOCK_FROM_SOURCE_LIBRARIES})
+      set(GMOCK_MAIN_LIBRARIES ${GMOCK_FROM_SOURCE_MAIN_LIBRARIES})
+      set(GMOCK_BOTH_LIBRARIES ${GMOCK_LIBRARIES} ${GMOCK_MAIN_LIBRARIES})
+
+      set(GTEST_FOUND ${GMOCK_FOUND})
+      set(GTEST_INCLUDE_DIRS ${GMOCK_INCLUDE_DIRS})
+      set(GTEST_LIBRARY_DIRS ${GMOCK_LIBRARY_DIRS})
+      set(GTEST_LIBRARIES ${GMOCK_LIBRARIES})
+      set(GTEST_MAIN_LIBRARIES ${GMOCK_MAIN_LIBRARIES})
+      set(GTEST_BOTH_LIBRARIES ${GMOCK_BOTH_LIBRARIES})
```

Comment:
Why are the gtest variables set to the same values as gmock here?

Same below.

Reply:
GMOCK_ variables include the GTEST headers and libraries as well.

The purpose of most of the code here, is to make sure that both gmock and gtest use the same version and are compiled together if they, that's why catkin_find_gmock_source calls for catkin_find_gtest_source.

Currently catkin_find_gmock_source does not export the GTEST_ variables to the parent scope, if this is needed I can add it.

Reply:
I guess the question (not yet asked / answered) is: how do we expect a user to use this?

Based on the current patch I assume that users should continue to call `catkin_add_gtest` / `catkin_add_executable_with_gtest` and are then able to use `gtest` as well as `gmock` (when available. Is that assumption correct?

I am not sure that is a good way. I would have expected that the existing functions just link against `gtest` (as they did before, without changing). And a new set of functions `catkin_add_gmock` / `catkin_add_executable_with_gmock` would be introduced which links against both `gmock` as well as `gtest`. A user then needs to opt-in explicitly by updating the called CMake function name to use `gmock`.

Reply:
Yes your understanding of this patch's behavior is correct.

I understand your point, for any pair of libraries that were unrelated I would completely agree. 
In this case gmock is an extension of gtest, that in theory should be harmless. So I would say that if keeping them together keeps things simple and doesn't introduce any side effects (that I've seen) it might be worth it.

Reply:
I am not use if that might not confuse users that `catkin_add_gtest` "sometimes" works with gmock and sometimes doesn't (e.g. when they use different versions of the ROS). ROS packages might start using `gmock` in their tests without realizing that the same API won't work in older ROS distros.

@clalancette @dhood @mikaelarguedas @tfoote @wjwwood Maybe any of you has feedback on this?

Reply:
+0

I don't have enough context to know if there is a better way to do this. (I might say +1 if I knew there was not for sure)

Developers might run into a mismatch in features (fails to compile in older distros after some changes), but if we document the potential change in behavior, then that's the best we can do. I don't think that risk is enough to avoid the improvement proposed.

Reply:
@wjwwood An option would be to introduce new CMake functions with `gmock` in the name. So users have to actively switch and those are not available in older distros.

Reply:
IMO there is value in keeping behavior consistent across distros and to let users opt-in to use new features. So I'd be in favor of having new cmake functions for users that want to use gmock. Then I don't have enough knowledge to estimate the risk and could be convinced that the risk is minimal enough for allowing a documented change of behavior

Reply:
Having new function/macro names sounds fine to me, I didn't realize that was an option. I thought this was a case where the new version of googletesting is just gmock and gtest isn't separate anymore. My confusion.

Reply:
Ok, I will add `catkin_add_gmock` and `catkin_add_executable_with_gmock`.

Before I continue with it, there will be a lot of similarities between the gtest and gmock functions.
My idea was to make the current 2 functions' code generic in tow new "private" functions, let's say `_catkin_add_google_test` and `_catkin_add_executable_with_google_test`, and have the gtest and gmock functions call this with minimum code duplication.

Would this be your preferred approach? It would make the code a little bit more complex, the other option is to duplicate and modify the code for the gmock versions, and leave the originals as they were.

Reply:
> ... with minimum code duplication.

@v-lopez That sounds good. Whatever avoids code duplication. Either some "private" functions which are used from the "public" ones. Or the `gmock` ones calling the `gtest` ones internally (if that is feasible).

Reply:
I've made the discussed changes in https://github.com/ros/catkin/pull/896/commits/3c4c7b2326fdec3f480fdc3a5934d59d7f05f1ea

There's one thing that I don't know how to test. All my changes are focused towards the gtest & gmock from sources approach, which as far as I know, is the recommended way of using google test.

But there's a `find_package(GMock QUIET)` and `find_package(GTest QUIET)` that I don't know how should behave, because there's not a unique FindGMock.cmake file, there are many implementations with different behaviours.